### DISCUSSION TITLE
Comparison of texture analysis/plot while importing orientation files as a generic and another after 'odf' calculation from orientation files

### DISCUSSION THREAD

**Discussion by AmitS07** [mtex-toolbox/mtex#1607] [Ask Anything]
Hello MTEX TEam and Community,

I have a doubt about plotting the texture (Pole Figure (PF), IPF, and ODF) from the as-imported orientations file. I did plot the texture by importing the orientation data as a generic (`loadOrientation_generic`) and plotted the PFs. Further, I calculated the odf form `calcDensity' and plotted the PFs. I found that MRD/Density was quite high (almost double) in the case of texture plotted from directly ori files (I mean loading orientation data as a generic). So, what is the cause of this? I did further plot IPF and ODF sections and found the same. I did a few exercises with 'random' and 'textured' data. I further, found out that there was much difference in the MRD/Density for the 'randomly' textured sample between both methods of plotting.  

I am attaching my results below with the MATLAB script.

Example 1:  I took a **'highly' texture** orientation file
(a) Plotted PF from ori files calculated form `loadOrientation_generic`
![1000_grain_text_fori_pf](https://user-images.githubusercontent.com/67289970/197589265-4a5be1e4-c7f0-4c81-a972-b3f22342c50c.png)
(b) Plotted PF after 'odf' calculation form `calcDensity`
![1000_grain_text_fodf_pf](https://user-images.githubusercontent.com/67289970/197589277-a9976b0f-762f-430e-bbf9-7be6e19b5364.png)
(c) Plotted IPF from ori files calculated form `loadOrientation_generic`
![1000_grain_text_fori_ipf](https://user-images.githubusercontent.com/67289970/197591833-355dac8a-830a-4fcb-a620-97ed84f74501.png)
(d) Plotted IPF after 'odf' calculation form `calcDensity`
![1000_grain_text_fodf_ipf](https://user-images.githubusercontent.com/67289970/197591842-50ed0543-7be9-4570-95c7-bd5d2dd69972.png)

Example 2:  I took a **'randomly' texture** orientation file
(a) Plotted PF from ori files calculated form `loadOrientation_generic`
![1000_grain_rand_fori_pf](https://user-images.githubusercontent.com/67289970/197590974-f4867895-3414-4b48-943b-67e344b643c3.png)
(b) Plotted PF after 'odf' calculation form `calcDensity`
![1000_grain_rand_fodf_pf](https://user-images.githubusercontent.com/67289970/197590996-ab745cbe-d837-452c-acdf-e5aabe24991d.png)
(c) Plotted IPF from ori files calculated form `loadOrientation_generic`
![1000_grain_rand_fori_ipf](https://user-images.githubusercontent.com/67289970/197592161-773394f4-a1e4-461b-8dba-c79952bcf39c.png)
(d) Plotted IPF after 'odf' calculation form `calcDensity`
![1000_grain_rand_fodf_ipf](https://user-images.githubusercontent.com/67289970/197592173-04441127-0c8a-43dc-b38b-d8df22bfa918.png)
Note: It is interesting to see that **'randomly' texture** data shows different orientations distributions between both PF/IPF plots i.e. between the data incorporated as generic and the other as odf.
Doubts:
1.  What causes the difference between both, since the raw data was the same? 
2. Which method of plotting texture can consider a correct one? 
I think since ODF calculation is led by some function, which may cause this difference.

I am also attaching my MATLAB script as well, kindly find it. 
```
%% code to plot Texture form orientation file%%
CS = crystalSymmetry('hexagonal');
SS = specimenSymmetry('monoclinic');
pf = readmatrix('caxis_1000rand.txt'); %pf = normalize(pf) : I think higher mrd might be due to normalization of the poles/orientations
ori = loadOrientation_generic(pf,'CS',CS,'SS',SS, 'ColumnNames', ...
{'Euler1' 'Euler2' 'Euler3'},'Columns',[1,2,3],'Bunge');
odf = calcDensity(ori);
% PF plot
h = Miller({0,0,0,1},ori.CS);
plotPDF(ori,h, 'projection', 'eangle','antipodal', 'false', 'contourf','complete', 'upper') % 'antipodal', 'silent', %'colorrange',[0 20],% 'complete', 'upper'
mtexColorMap('viridis')
mtexColorbar  %plotPDF(ori,h,'3d', 'points', 'all') % to plot in 3D
figure()
plotPDF(odf,h, 'projection', 'eangle','antipodal', 'false','contourf', 'complete', 'upper')  % 'colorRange','equal' for only on ecolor map
mtexColorMap('viridis')
mtexColorbar
figure()
% IPF plot
plotIPDF(ori,zvector,'contourf','antipodal','false')
mtexColorMap('viridis')
mtexColorbar
figure()
plotIPDF(odf,zvector,'contourf','antipodal','false');
mtexColorMap('viridis')
mtexColorbar
figure()
% ODF plot
plot(odf,'rodrigues')
mtexColorMap('viridis')
figure()
% figure()
% plot(odf,'phi2',60*degree)
% mtexColorMap('viridis')
```

[pf_analysis_caxis.txt](https://github.com/mtex-toolbox/mtex/files/9854163/pf_analysis_caxis.txt)

**Comment by zmichels** [comment]
Sometimes for me I find I have to specify the halfwidth during the construction of the ODF… because attempts to specify during plotPDF() don’t seem to override whatever was used in calcDensity(). If you specify your halfwidth during construction… does that change your plots?ZachOn Oct 24, 2022, at 11:10 AM, Amit Singh ***@***.***> wrote:﻿
Hello MTEX TEam and Community,
I have a doubt about plotting the texture (Pole Figure (PF), IPF, and ODF) from the as-imported orientations file. I did plot the texture by importing the orientation data as a generic (loadOrientation_generic) and plotted the PFs. Further, I calculated the odf form `calcDensity' and plotted the PFs. I found that MRD/Density was quite high (almost double) in the case of texture plotted from directly ori files (I mean loading orientation data as a generic). So, what is the cause of this? I did further plot IPF and ODF sections and found the same. I did a few exercises with 'random' and 'textured' data. I further, found out that there was much difference in the MRD/Density for the 'randomly' textured sample between both methods of plotting.
I am attaching my results below with the MATLAB script.
Example 1:  I took a 'highly' texture orientation file
(a) Plotted PF from ori files calculated form loadOrientation_generic

(b) Plotted PF after 'odf' calculation form calcDensity

(c) Plotted IPF from ori files calculated form loadOrientation_generic

(d) Plotted IPF after 'odf' calculation form calcDensity

Example 2:  I took a 'randomly' texture orientation file
(a) Plotted PF from ori files calculated form loadOrientation_generic

(b) Plotted PF after 'odf' calculation form calcDensity

(c) Plotted IPF from ori files calculated form loadOrientation_generic

(d) Plotted IPF after 'odf' calculation form calcDensity

Note: It is interesting to see that 'randomly' texture data shows different orientations distributions between both PF/IPF plots i.e. between the data incorporated as generic and the other as odf.
Doubts:

What causes the difference between both, since the raw data was the same?
Which method of plotting texture can consider a correct one?
I think since ODF calculation is led by some function, which may cause this difference.

I am also attaching my MATLAB script as well, kindly find it.
%% code to plot Texture form orientation file%% CS = crystalSymmetry('hexagonal'); SS = specimenSymmetry('monoclinic'); pf = readmatrix('caxis_1000rand.txt'); %pf = normalize(pf) : I think higher mrd might be due to normalization of the poles/orientations ori = loadOrientation_generic(pf,'CS',CS,'SS',SS, 'ColumnNames', ... {'Euler1' 'Euler2' 'Euler3'},'Columns',[1,2,3],'Bunge'); odf = calcDensity(ori); % PF plot h = Miller({0,0,0,1},ori.CS); plotPDF(ori,h, 'projection', 'eangle','antipodal', 'false', 'contourf','complete', 'upper') % 'antipodal', 'silent', %'colorrange',[0 20],% 'complete', 'upper' mtexColorMap('viridis') mtexColorbar  %plotPDF(ori,h,'3d', 'points', 'all') % to plot in 3D figure() plotPDF(odf,h, 'projection', 'eangle','antipodal', 'false','contourf', 'complete', 'upper')  % 'colorRange','equal' for only on ecolor map mtexColorMap('viridis') mtexColorbar figure() % IPF plot plotIPDF(ori,zvector,'contourf','antipodal','false') mtexColorMap('viridis') mtexColorbar figure() plotIPDF(odf,zvector,'contourf','antipodal','false'); mtexColorMap('viridis') mtexColorbar figure() % ODF plot plot(odf,'rodrigues') mtexColorMap('viridis') figure() % figure() % plot(odf,'phi2',60*degree) % mtexColorMap('viridis')
pf_analysis_caxis.txt

**Reply by AmitS07** [reply]
Hi Michels, 

Thank you for the reply. I did not apply `halfwidth` during the construction of ODF. Maybe I can try, but exactly what `halfwidth' one has to take? I checked in MTEX documents it is difficult to estimate. However, I will try to do 'optimal Kernel calculation then I can try to recalculate ODF. Thank you for the suggestion. I will update, with the results.

**Comment by zmichels** [comment]
Probably this is not the issue… but I guess my point was to try specifying the same halfwidth in both plots (and maybe have to do it at the construction, too, for the ODF)… to be sure that both plots are using the same halfwidth for MRD.ZachOn Oct 24, 2022, at 12:46 PM, Amit Singh ***@***.***> wrote:﻿
Hi Michels,
Thank you for the reply. I did not apply halfwidth' during the construction of ODF. Maybe I can try, but exactly what halfwidth' one has to take? I checked in MTEX documents it is difficult to estimate. However, I will try to do 'optimal Kernel calculation then I can try to recalculate ODF. Thank you for the suggestion. I will update, with the results.

**Reply by AmitS07** [reply]
Hi, 
if  I am calculating `ori` with `halfwidth`  as following, then there is no change between both.  The MRD of PF plotted after odf calculation is remain 25, and with `generic' one remains ~60. 
```
ori = loadOrientation_generic(pf,'CS',CS,'SS',SS, 'ColumnNames', ...
{'Euler1' 'Euler2' 'Euler3'},'Columns',[1,2,3],'Bunge', 'halfwidth',10*degree);
```
However, MRD changes while incorporating optimal Kernel, for which the `halfwidth` is 3 and `bandwidth` 78 degree. The MRD reached 70. The MRD does not get affected by the ` 'halfwidth',10*degree`. I did remove this and reported, I got the same plot with the same MRD.
```
ori = loadOrientation_generic(pf,'CS',CS,'SS',SS, 'ColumnNames', ...
{'Euler1' 'Euler2' 'Euler3'},'Columns',[1,2,3],'Bunge', 'halfwidth',10*degree);
psi = calcKernel(ori);
odf = calcDensity(ori,'kernel',psi);
```
![with_kernel](https://user-images.githubusercontent.com/67289970/197639582-c68783e9-0a00-470b-853f-ad386f0c7eee.png)

**Comment by ralfHielscher** [comment]
Please have a look at https://mtex-toolbox.github.io/DensityEstimation.html
The point is that there is not "one" density distribution for a given set of orientations. You have to select the most reasonable using the option "halfwidth".

Ralf.

**Reply by AmitS07** [reply]
Hello Dr. Hielscher, 

Thank you for the reply. 
So, I tried with `halfwidth`, but did not find much difference. How one can optimize the `halfwidth` value?

What is the better way to plot the PF, I mean by considering `halwidth` or without. What is the role of `kernel'.
I have done some exercises and the results are below. 
Please suggest to me the appropriate method to plot the PFs.
![Picture2](https://user-images.githubusercontent.com/67289970/199808712-18a174dc-be8c-4eb9-99cf-5ac412aa9a3d.png)

NOTE: As can be seen that there are huge differences in the MRD of PF plots from different methods.
The code is here:
`CS = crystalSymmetry('hexagonal');`
`SS = specimenSymmetry('monoclinic');`
`pf = readmatrix('caxis_1000text.txt'); `
`ori = loadOrientation_generic(pf,'CS',CS,'SS',SS, 'ColumnNames', ...
{'Euler1' 'Euler2' 'Euler3'},'Columns',[1,2,3],'Bunge'); %, 'halfwidth',10*degree`
`%psi = calcKernel(ori);` % this lines **uncommented** while calculating odf from kernel
`%odf = calcDensity(ori,'kernel',psi);` % this lines **uncommented** while calculating odf from kernel
`odf = calcDensity(ori);` % this line **commented**  while calculating odf from kernel
`h = Miller({0,0,0,1},ori.CS);`
`plotPDF(ori,h, 'projection', 'eangle','antipodal', 'false', 'contourf','complete', 'upper')`
`mtexColorMap('viridis')`
`mtexColorbar `
`figure()`
`plotPDF(odf,h, 'projection', 'eangle','antipodal', 'false','contourf','complete', 'upper')`
`mtexColorMap('viridis')`
`mtexColorbar`
`figure()`
Q. Why there are huge differences in MRDs? what is the correct way to plot PF?

**Reply by ralfHielscher** [reply]
Hi Amit,

(1) the ```halfwidth``` should be passed to the calcDensity function not to the ```load``` function.
(2) whenever, you do not pass a halfwidth (or equivalently a kernel) MTEX assumes some default value. 
(3) I completely understand your frustration, but if other software suggests it can plot a single pole figure to a set of orientations - it is not telling the truth. Selecting the most reasonable halfwidth (there is no correct halfwidth) is a difficult problem. the function ```calcKernel``` attempts to do this for you. You can either trust this function or specify your own halfwidth. 

Ralf.

**Reply by AmitS07** [reply]
Hello Dr. Ralf, 

Thank you very much for the explanation and suggestions.  Now, I can better understand the 'halfwidth'. I will do few exercises and try to optimize the 'halfwidth'.

**Comment by AmitS07** [comment]
Hello all, 

Could anyone please suggest the correct way of plotting the pole figure? Code that I used have used is mentioned above.
As it can be seen from the above plot that all three pole figures show the same, texture. But different mrd/density varies based on the method used for pole figure calculation. So, what is the correct way of PF calculation?
